lapack/gonum/dlatbs.go

```diff
@@ -0,0 +1,449 @@
+			//
+			// Compute grow = 1/G_j and xBnd = 1/M_j.
+			// Initially, G_0 = max{x(i), i=1,...,n}.
+			grow = 1 / math.Max(xBnd, smlnum)
+			xBnd = grow
+			for j := jFirst; j != jLast; j += jInc {
+				if grow <= smlnum {
+					// Exit the loop because the growth factor is too small.
+					goto skipComputeGrow
+				}
+				// G_j = max( G_{j-1}, M_{j-1}*( 1 + cnorm[j] ) )
+				xj := 1 + cnorm[j]
+				grow = math.Min(grow, xBnd/xj)
+				// M_j = M_{j-1}*( 1 + cnorm[j] ) / abs(A[j,j])
+				tjj := math.Abs(ab[j*ldab+maind])
+				if xj > tjj {
+					xBnd *= tjj / xj
+				}
+			}
+			grow = math.Min(grow, xBnd)
+		} else {
+			// A is unit triangular.
+			//
+			// Compute grow = 1/G_j, where G_0 = max{x(i), i=1,...,n}.
+			grow = math.Min(1, 1/math.Max(xBnd, smlnum))
+			for j := jFirst; j != jLast; j += jInc {
+				if grow <= smlnum {
+					// Exit the loop because the growth factor is too small.
+					goto skipComputeGrow
+				}
+				// G_j = G_{j-1}*( 1 + cnorm[j] )
+				grow /= 1 + cnorm[j]
+			}
+		}
+	}
+skipComputeGrow:
+
+	if grow*tscal > smlnum {
+		// The reciprocal of the bound on elements of X is not too small, use
+		// the Level 2 BLAS solve.
+		bi.Dtbsv(uplo, trans, diag, n, kd, ab, ldab, x, 1)
+		// Scale the column norms by 1/tscal for return.
+		if tscal != 1 {
+			bi.Dscal(n, 1/tscal, cnorm, 1)
+		}
+		return 1
+	}
+
+	// Use a Level 1 BLAS solve, scaling intermediate results.
+
+	scale = 1
+	if xMax > bignum {
+		// Scale x so that its components are less than or equal to bignum in
+		// absolute value.
+		scale = bignum / xMax
+		bi.Dscal(n, scale, x, 1)
+		xMax = bignum
+	}
+
+	if noTran {
+		// Solve A * x = b.
+		for j := jFirst; j != jLast; j += jInc {
+			// Compute x[j] = b[j] / A[j,j], scaling x if necessary.
+			xj := math.Abs(x[j])
+			tjjs := tscal
+			if diag == blas.NonUnit {
+				tjjs *= ab[j*ldab+maind]
+			}
+			tjj := math.Abs(tjjs)
+			switch {
+			case tjj > smlnum:
+				// smlnum < abs(A[j,j])
+				if tjj < 1 && xj > tjj*bignum {
+					// Scale x by 1/b[j].
+					rec := 1 / xj
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+					xMax *= rec
+				}
+				x[j] /= tjjs
+				xj = math.Abs(x[j])
+			case tjj > 0:
+				// 0 < abs(A[j,j]) <= smlnum
+				if xj > tjj*bignum {
+					// Scale x by (1/abs(x[j]))*abs(A[j,j])*bignum to avoid
+					// overflow when dividing by A[j,j].
+					rec := tjj * bignum / xj
+					if cnorm[j] > 1 {
+						// Scale by 1/cnorm[j] to avoid overflow when
+						// multiplying x[j] times column j.
+						rec /= cnorm[j]
+					}
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+					xMax *= rec
+				}
+				x[j] /= tjjs
+				xj = math.Abs(x[j])
+			default:
+				// A[j,j] == 0: Set x[0:n] = 0, x[j] = 1, and scale = 0, and
+				// compute a solution to A*x = 0.
+				for i := range x[:n] {
+					x[i] = 0
+				}
+				x[j] = 1
+				xj = 1
+				scale = 0
+				xMax = 0
+			}
+
+			// Scale x if necessary to avoid overflow when adding a multiple of
+			// column j of A.
+			if xj > 1 {
```

Comment:
`switch` since this seems to be a switch-prone function?

Reply:
Ok, will do.

Reply:
Done.